### PR TITLE
New union to help packing and unpacking of data

### PR DESCRIPTION
**Pull request by indraneelpatil**
## Description

- Added union for message building
- Replaced message handling functions with a generic handling function
- Added test file for the above change

## 

## How has this been tested?

Tested against the test data file on the branch 

Message Name: MOONRANGER_WheelVelocity_Command_t
Message Len: MOONRANGER_WHEEL_VEL_CMD_MSG_LNGTH (this is a constant that is declared)
Message ID: 0x0D09 
Message Data:
    Duration - 10
    leftFront - 8.0
    rightFront - 4.0
    leftBack - 6.0 
    rightBack - 2.0`

`Hex encoding of CCSDS Packed Message (this is the payload of the UDP packet)
0d05c000001d0000000000000a00000000000041000080400000c04000000040`

**Testing Output : Extracting**
10
8.000000
4.000000
6.000000
2.000000

**Testing Output : Building**
d 5 c0 0 0 19 0 0 0 0 0 0 a 0 0 0 0 0 0 41 0 0 80 40 0 0 c0 40 0 0 0 40

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] I have read and followed the `CONTRIBUTING.md` file
- [ ] I have applied the .clang-format file to my changes
- [ ] I have commented my code using Doxygen style comments (see Contributing MD for examples), particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the design documentation 
- [ ] I have updated any relevant parameters in the internal ICD
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If this is a bug-fix, I have added tests that can catch similar bugs in the future
- [ ] New and existing unit tests pass locally with my changes
- [ ] For cfe apps I have reserved an empty address for my msgs_ids and perf_ids from the [tracking list](https://docs.google.com/spreadsheets/d/1yRUhNTRBOiGLswlsWftNaYlwGKDeNQ_fCnDYFrc_7fU/edit#gid=772812381) and checked that it does not clash with other apps
- [ ] For new added message types, there is a timestamp included via either an explicit timestamp variable in the struct or buried in the a tlmHeader.

### CODE REVIEW

File: fsw/public_inc/pose_msg.h [hunk 1]
```diff
@@ -21,7 +21,7 @@
 /*
 ** Type definition (MOONRANGER pose)
 */
-typedef struct {
+typedef struct OS_PACK {
```
Comment (tenzincrouch) [inline fsw/public_inc/pose_msg.h:24]: Is this to ensure word alignment for every data struct?

File: message-utilites/cfe_components/cfe_time.h [hunk 2]
```diff
@@ -112,8 +112,7 @@
 **       elapsed since the epoch.
 **
 */
-typedef struct
-{
+typedef struct OS_PACK  {
```
Comment (tenzincrouch) [inline message-utilites/cfe_components/cfe_time.h:115]: We may have to adjust this on the flight software too. This file is used by both.

File: message-utilites/src/message_builder.c [hunk 3]
```diff
@@ -15,47 +15,43 @@
...
-}
+  // Fill the data
 
-void BuildTeleopCmd(void *MsgPtr, int32 msgId, MOONRANGER_Teleop_t *data){
-  MOONRANGER_Teleop_Cmd_t *ptr = MsgPtr;
-  CFE_SB_InitMsg(ptr, (CFE_SB_MsgId_t)msgId, MOONRANGER_TELEOP_CMD_LNGTH, true);
+  // Sanity check what we are about to do
+  if(data_len_bytes+header_length<=sizeof(message_builder_u))
+  {
+    memcpy(&msg_container->data_buf_ptr,DataPtr,data_len_bytes);
```
Comment (tenzincrouch) [inline message-utilites/src/message_builder.c:51]: The one concern I have here is whether we care about zero copy or not? I think the message volume and size is low enough that we dont care?

**Review by tenzincrouch — COMMENTED** [review]
I think this is good. Endianness should not affect this because it is only serialized / deserialised in the actual send /receive not pack.

What testing have you done with this?

File: fsw/public_inc/pose_msg.h (hunk 1 above)
Comment (indraneelpatil) [reply fsw/public_inc/pose_msg.h:24]: Yes

File: message-utilites/src/message_builder.c (hunk 3 above)
Comment (indraneelpatil) [reply message-utilites/src/message_builder.c]: Need to give this more thought, Message building is zero copy, message extracting uses a memcopy, looks difficult to implement a zero copy implementation without breaking a lot of compatibility

FYI lwip on PC size is also not zero copy

**Comment by indraneelpatil** [comment]
@tenzincrouch  Added a unit test for this https://github.com/PlanetaryRobotics/moonranger-rover-messages/pull/22/commits/24ca38cdad0a74f33bd70a7ba4f784834c480fb8

Please see the 'How has this been tested section'

**Comment by indraneelpatil** [comment]
@tenzincrouch In the test data that you have attached, why are the length bytes 001d? 1d is 29 bytes but the size of the data is 32 bytes, thats the only byte which does not match

**Review by tenzincrouch — APPROVED** [review]

**Pull request merged**